coordinator/src/main/scala/filodb.coordinator/NodeClusterActor.scala

```diff
@@ -0,0 +1,99 @@
+package filodb.coordinator
+
+import akka.actor.{ActorRef, Props, RootActorPath}
+import akka.cluster.Cluster
+import akka.cluster.ClusterEvent._
+import akka.event.LoggingReceive
+import scala.collection.mutable.HashMap
+import scala.concurrent.duration._
+
+import filodb.core.ErrorResponse
+
+object NodeClusterActor {
+  // Forwards message to one random recipient that has the given role.  Any replies go back to originator.
+  case class ForwardToOne(role: String, msg: Any)
+  case object NoSuchRole extends ErrorResponse
+
+  // Gets all the ActorRefs for a specific role.  Returns a Set[ActorRef].
+  case class GetRefs(role: String)
+
+  // Forwards message to all recipients with given role.  Sending actor must handle separate replies.
+  case class ForwardToAll(role: String, msg: Any)
+
+  // Internal message
+  case class AddCoordActor(roles: Set[String], ref: ActorRef)
+
+  def props(cluster: Cluster,
+            resolveActorTimeout: FiniteDuration = 10.seconds): Props =
+    Props(classOf[NodeClusterActor], cluster, resolveActorTimeout)
+}
+
+/**
+ * An actor that subscribes to membership events for a FiloDB cluster, updating a state of coordinators.
+ * It also can send messages to one or all coordinators of a given role.
+ *
+ * Compared to the standard cluster aware routers, it has a couple advantages:
+ * - It tracks all roles, not just one, and can send messages to a specific role per invocation
+ * - It has much lower latency - direct sending to ActorRefs, no need to resolve ActorSelections
+ *   (but of course it assumes the destination actor is always up... TODO watch for other actor dying)
+ * - It broadcasts messages to all members of a role and sends back a collected response in one message,
+ *   making it very easy to handle for non-actors.
+ * - It can notify when some address joins
+ */
+class NodeClusterActor(cluster: Cluster,
+                       resolveActorTimeout: FiniteDuration) extends BaseActor {
+  import NodeClusterActor._
+
+  val roleToCoords = new HashMap[String, Set[ActorRef]]().withDefaultValue(Set.empty[ActorRef])
+
+  import context.dispatcher
+
+  // subscribe to cluster changes, re-subscribe when restart
+  override def preStart(): Unit = {
+    cluster.subscribe(self, initialStateMode = InitialStateAsEvents,
+      classOf[MemberEvent], classOf[UnreachableMember])
+  }
+
+  // TODO: leave cluster?
+  override def postStop(): Unit = cluster.unsubscribe(self)
+
+  private def withRole(role: String)(f: Set[ActorRef] => Unit): Unit = roleToCoords.get(role) match {
+    case None       => sender ! NoSuchRole
+    case Some(refs) => f(refs)
+  }
+
+  def receive: Receive = LoggingReceive {
+    case MemberUp(member) =>
+      logger.info(s"Member is Up: ${member.address} with roles ${member.roles}")
```

Comment:
I think it doesn’t log the roles though.

> On May 5, 2016, at 10:26 AM, Helena Edelson <notifications@github.com> wrote:
> 
> In coordinator/src/main/scala/filodb.coordinator/NodeClusterActor.scala <https://github.com/tuplejump/FiloDB/pull/94#discussion_r62221506>:
> 
> > +  override def preStart(): Unit = {
> > +    cluster.subscribe(self, initialStateMode = InitialStateAsEvents,
> > +      classOf[MemberEvent], classOf[UnreachableMember])
> > +  }
> > +
> > +  // TODO: leave cluster?
> > +  override def postStop(): Unit = cluster.unsubscribe(self)
> > +
> > +  private def withRole(role: String)(f: Set[ActorRef] => Unit): Unit = roleToCoords.get(role) match {
> > +    case None       => sender ! NoSuchRole
> > +    case Some(refs) => f(refs)
> > +  }
> > +
> > +  def receive: Receive = LoggingReceive {
> > +    case MemberUp(member) =>
> > +      logger.info(s"Member is Up: ${member.address} with roles ${member.roles}")
> usually akka cluster logs this already, unless it's disabled
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub <https://github.com/tuplejump/FiloDB/pull/94/files/94fb535ac49b74f362beb550bf1989a3ad04d3e3#r62221506>